KIALI-1810 Introduce first level cache for k8s API

lucasponce opened the pull request:
This is a first step to use client-go/cache cappabilities inside kubernetes package.
The use of this cache is part of the implementation of the kubernetes package, it is transparent to the other layers of the architecture.
Internally, Informers allow to list and watch k8s objects, providing default indexing for namespace and names which is the most of our use cases.
On this way, backend should improve the performance globally (specially on the heaviest operations, as identify workloads).

Please @jotak and @hhovsepy take a close look at it, I am curious to see how it works in the QE environment as well.

rhqci commented:
:heavy_check_mark: Jenkins CI: kiali-core-pr-e2e-test [#250](https://paste.fedoraproject.org/paste/AFwBIuPGI3kieUI2CuwOlQ)
  * :red_circle: Istio-and-Kiali-on-Openshift [#[1758]](https://paste.fedoraproject.org/paste/GOXxDubYtgoXRDhKvKKWlQ)
  * :heavy_check_mark: Istio-Upstream [#[604]](https://paste.fedoraproject.org/paste/cXT-jMMVRlpCLoh6knDqew)
  * :heavy_check_mark: Kiali-Istio-Checker [#[2064, 2063]](https://paste.fedoraproject.org/paste/WsjfgtGjq5Fn7ssT-ZFalw)
  * :heavy_check_mark: install-bookinfo [#[1134]](https://paste.fedoraproject.org/paste/AGqGbTFAnjCuVCdhHsn5~w)
  * :heavy_check_mark: Kiali-on-Openshift [#[1124]](https://paste.fedoraproject.org/paste/q5LiZWV2SQfcGW1Wfz7C9w)
  * :red_circle: run-kiali-e2e-tests [#[530]](https://paste.fedoraproject.org/paste/YApKU5mTX6gYbWWTeoPjjA)


KIALI_NAME=kialicorepr588-istio-system

jotak commented on kubernetes/kubernetes_service.go:
```diff
@@ -135,94 +140,87 @@ func (in *IstioClient) GetDeploymentConfigs(namespace string) ([]osappsv1.Deploy
+	if labelSelector != "" {
+		selector, err := labels.Parse(labelSelector)
+		if err != nil {
+			return pods, err
```
I think it would be safer to return an empty array here, to avoid the risk that the caller takes the list of pods ignoring the error.

lucasponce replied:
+1

jotak replied:
don't forget this one :)

lucasponce replied:
Oh, true thanks ! I'm on my way, having issues with the local test env this morning, but I'll update it

jotak commented on kubernetes/client.go:
```diff
@@ -79,6 +79,10 @@ type IstioClient struct {
 	istioConfigApi     *rest.RESTClient
 	istioNetworkingApi *rest.RESTClient
+	// isOpenShift can be cached per client
+	isOpenShift *bool
```
could it be just "bool" ? (i don't know, it seems weird to get a pointer just for a boolean - what's the reason?)

lucasponce replied:
I wanted to represent 3 states: not defined, true or false.
So, if it's not defined, it means that kiali didn't query if we are in kubernetes or openshift.
Returning a false in all cases could be misleading.

jotak replied:
ah, ok

jshaughn replied:
although, it seems more typical in Go to return two values in that case.

jmazzitelli replied:
I agree that Go typically returns two values for this kind of thing (IsOpenShift() (bool err)).
If this code remains using a pointer, I highly recommend you put a comment above this to explain why it is a pointer and that a nil is expected when not set.

lucasponce replied:
This code is going to use a pointer as it represents three states and no two.

jotak commented on kubernetes/cache.go:
```diff
@@ -0,0 +1,325 @@
+func (c *controllerImpl) GetServices(namespace string) ([]v1.Service, bool) {
+	indexer := c.controllers["Service"].GetIndexer()
+	services, err := indexer.ByIndex("namespace", namespace)
+	if err != nil {
```
what kind of error could we have here?
maybe it's worth logging, and even sending as a result?

lucasponce replied:
I'm a going to check.
I did some tests to validate the recovery cappabilities (error propagation when backend is off, reconnect, etc) but I don't remember if this is missing.
I will update my comment after a couple of tests.

jotak commented on kubernetes/cache.go:
```diff
@@ -0,0 +1,325 @@
+	}
+	if len(services) > 0 {
+		_, ok := services[0].(*v1.Service)
+		if !ok {
```
If I understand correctly, this is to check that the cache contains objects of the expected type?
So, if not, maybe we should propagate some kind of "Internal Error", no? (and log)

lucasponce replied:
+1

jotak commented on kubernetes/cache.go:
```diff
@@ -0,0 +1,325 @@
+
+func (c *controllerImpl) GetService(namespace, name string) (*v1.Service, bool) {
+	indexer := c.controllers["Service"].GetIndexer()
+	services, exist, err := indexer.GetByKey(namespace + "/" + name)
```
Trying to understand how the cache works... is `GetByKey` as fast as `ByIndex` (in other words, keys are also indexed? it's not going to perform full search / O(N) ?)

lucasponce replied:
GetByKey() is O(1) as it fetches from the map.
Internally, the cache uses as native keys the same "namespace/name" pattern.
So, before I added a similar index, but then I realized that was not necessary as I can use the native key.

jotak commented on kubernetes/cache.go:
```diff
@@ -0,0 +1,325 @@
+
+func (c *controllerImpl) GetDeployment(namespace, name string) (*v1beta1.Deployment, bool) {
+	indexer := c.controllers["Deployment"].GetIndexer()
+	deps, exist, err := indexer.GetByKey(namespace + "/" + name) // indexer.ByIndex("nsname", namespace + "#" + name)
```
What's the commented out code here? Another way to get same result, or just a relic of your investigations? :)

lucasponce replied:
Ah, left code there ! Forgot to clean, perhaps there are a couple of unnecessary logging out there.
Thanks

jotak commented on kubernetes/cache.go:
```diff
@@ -0,0 +1,325 @@
+func (c *controllerImpl) HasSynced() bool {
+	if c.syncCount > c.maxSyncCount {
+		log.Errorf("Max attempts reached syncing cache. Is there connection with the K8S backend ?")
+		c.StopControlChannel()
```
what happens here exactly? Are we entering in an unrecoverable state by stopping control channel?

lucasponce replied:
On initialization, I see usual to have an active wait until cache is populated:
```
cache.WaitForCacheSync(client.stopCache, client.k8sCache.HasSynced)
```
(I saw this on tests and also in usage of the cache)

But it happens that there is a cornercase when the backend is always down, so, to resolve this properly I forced to fail.

In the preliminar testing I did, it looked good.
But with your comments and also other ideas I thought I guess cache could hide better the stopchannel and provide a simple way for this; I'm refactoring that part, I'll update the PR later.

jotak commented:
Great PR!
Several questions inline.

Also, a question that I have regarding the use of cache in general: a side effect is that the user will be less informed when something's wrong when we try to access the k8s API. There's no more synchronous errors from k8s client when we call, for instance, `GetServices`, we're kind of turning it into a black box.

How could we tackle that? The information of connection failures should appear up to the UI.
Maybe it's the right time to start this task: https://issues.jboss.org/browse/KIALI-1118

jotak commented:
I've tested with this PR:
- Running Kiali without any service mesh deployed => everything looks fine in Overview, Apps, Workloads, Services
- Adding a mesh with deployments and services => everything's fine, added entities are displayed
- Removing one deployment and service => correctly updated
- Increasing number of replica => number of pods in workload details correctly updated
- Modifying a pod labels => pod labels correctly updated in the UI

So, everything looks good so far!

jotak commented:
Hmm some comments about the server logs:
- watchers are continuously logging some information such as:
`I1026 08:10:45.615540       1 reflector.go:428] github.com/kiali/kiali/kubernetes/cache.go:73: Watch close - *v1beta2.ReplicaSet total 0 items received`
Maybe we should try to prevent that kind of information.

- One suspicious log entry comes regularly:
`W1026 08:12:03.083442       1 reflector.go:341] github.com/kiali/kiali/kubernetes/cache.go:73: watch of *v1beta1.Deployment ended with: The resourceVersion for the provided watch is too old.`
Something wrong here, I guess?

lucasponce commented:
Thanks for your review @jotak, I'm going to address your comments.
@hhovsepy could you test it ? focus on regressions and performance comparison ? Thanks

jshaughn commented:
@lucasponce Any idea of the perf impact?

jmazzitelli commented:
> @lucasponce Any idea of the perf impact?

We'll see once we add metrics around this stuff ;)

lucasponce commented:
> @lucasponce Any idea of the perf impact?

Locally it looks promising, but I asked @hhovsepy to run more formal testing around it.
The PR is not ready but it can be tested, it is a first step but it should have a positive impact on all K8S APIs.

So, graph appenders, health and app/workload/service list should get a better performance in general.

hhovsepy commented:
kialiqeci

rhqci commented:
:heavy_check_mark: Jenkins CI: kiali-core-pr-e2e-test [#263](https://paste.fedoraproject.org/paste/KLt-MlYk06FXg6PkeASBXA)
  * :red_circle: Istio-and-Kiali-on-Openshift [#[1783]](https://paste.fedoraproject.org/paste/8blCQoXZBknzdIpG6HK1CQ)
  * :heavy_check_mark: Istio-Upstream [#[628]](https://paste.fedoraproject.org/paste/e~Wxriaf7Th-DEugWPenrw)
  * :heavy_check_mark: Kiali-Istio-Checker [#[2120, 2119]](https://paste.fedoraproject.org/paste/eMObRHATI4fuOG5dgHzIbw)
  * :heavy_check_mark: install-bookinfo [#[1160]](https://paste.fedoraproject.org/paste/WOv1Q~TX6E4Zd2eFKzCbmA)
  * :heavy_check_mark: Kiali-on-Openshift [#[1158]](https://paste.fedoraproject.org/paste/JrK3nCchw3PJ43DkeDR~jA)
  * :red_circle: run-kiali-e2e-tests [#[562]](https://paste.fedoraproject.org/paste/cr8utR9agH4CW2pAafHsQw)


KIALI_NAME=kialicorepr588-istio-system

lucasponce commented:
@jotak I've sent a commit addressing some of your comments:
- Cache controller are part of the implementation and API is simplified
```
// Control Cache
Start()
HasSynced() bool
WaitForSync() bool
Stop()
```
- Cache business methods returns proper empty arrays and error in API
```
// Business methods
GetCronJobs(namespace string) ([]batch_v1beta1.CronJob, error)
GetDeployment(namespace string, name string) (*v1beta1.Deployment, error)
GetDeployments(namespace string) ([]v1beta1.Deployment, error)
GetEndpoints(namespace, name string) (*v1.Endpoints, error)
GetJobs(namespace string) ([]batch_v1.Job, error)
GetPods(namespace string) ([]v1.Pod, error)
GetReplicationControllers(namespace string) ([]v1.ReplicationController, error)
GetReplicaSets(namespace string) ([]v1beta2.ReplicaSet, error)
GetService(namespace string, name string) (*v1.Service, error)
GetServices(namespace string) ([]v1.Service, error)
GetStatefulSet(namespace string, name string) (*v1beta2.StatefulSet, error)
GetStatefulSets(namespace string) ([]v1beta2.StatefulSet, error)
```
- When cache can't sync with backend, cache stops and error is propagated; by default, cache is designed out of the box to retrying to connect in a transparent way; with this approach we maintain the feature we have today: to propagate in the UI a problem with the backend, but the cache is still working.

Probably there should be some adjustment more.

rhqci commented:
:heavy_check_mark: Jenkins CI: kiali-core-pr-e2e-test [#267](https://paste.fedoraproject.org/paste/V-~6x4eKtTEOi~jYHVfJYA)
  * :red_circle: Istio-and-Kiali-on-Openshift [#[1788]](https://paste.fedoraproject.org/paste/QgAwNVmBkg0s9dHIoUSchA)
  * :heavy_check_mark: Istio-Upstream [#[632]](https://paste.fedoraproject.org/paste/oUNQivpfGLKRq50BqKQVjw)
  * :heavy_check_mark: Kiali-Istio-Checker [#[2128, 2129]](https://paste.fedoraproject.org/paste/oHzXLx-W~CjvBilePT3oAg)
  * :heavy_check_mark: install-bookinfo [#[1165]](https://paste.fedoraproject.org/paste/UEjSuBwyRtIdWUm3m~xlvA)
  * :heavy_check_mark: Kiali-on-Openshift [#[1163]](https://paste.fedoraproject.org/paste/CZemHINVKWH2fdaX8wPntA)
  * :red_circle: run-kiali-e2e-tests [#[567]](https://paste.fedoraproject.org/paste/NNVakvYjuKt4kWDdBVbbWA)


KIALI_NAME=kialicorepr588-istio-system

rhqci commented:
:heavy_check_mark: Jenkins CI: kiali-core-pr-e2e-test [#268](https://paste.fedoraproject.org/paste/5yyA~UoojtM7zDSQo62AaA)
  * :red_circle: Istio-and-Kiali-on-Openshift [#[1789]](https://paste.fedoraproject.org/paste/u4qJKJ3YgXD4GhwWEdVzIw)
  * :heavy_check_mark: Istio-Upstream [#[633]](https://paste.fedoraproject.org/paste/QIEZyimD-BT4dW-MzVKurw)
  * :heavy_check_mark: Kiali-Istio-Checker [#[2130, 2131]](https://paste.fedoraproject.org/paste/Kx91CTpHSmRkEvRbI6fjkw)
  * :heavy_check_mark: install-bookinfo [#[1166]](https://paste.fedoraproject.org/paste/5yX-FzFAwUd6mm26yvco8A)
  * :heavy_check_mark: Kiali-on-Openshift [#[1164]](https://paste.fedoraproject.org/paste/pgXvYtS5TfDd5U6l1p45UA)
  * :red_circle: run-kiali-e2e-tests [#[568]](https://paste.fedoraproject.org/paste/bJrlYP-qicqxmQvoRhijyw)


KIALI_NAME=kialicorepr588-istio-system

lucasponce commented:
Using a proxy to connect with the API showed different results compared when connection is inside cluster.
I see a high level of `Watch close` messages that didn't appear with the former testing.
So, I'm still learning why this is happening.

rhqci commented:
:heavy_check_mark: Jenkins CI: kiali-core-pr-e2e-test [#272](https://paste.fedoraproject.org/paste/rbY-4~dDK1Dbr4sPlpoOew)
  * :red_circle: Istio-and-Kiali-on-Openshift [#[1797]](https://paste.fedoraproject.org/paste/oPhqtqz0ONJX3dR5EirkhQ)
  * :heavy_check_mark: Istio-Upstream [#[640]](https://paste.fedoraproject.org/paste/-LZyYe3AW9jDeOD5VubtEg)
  * :heavy_check_mark: Kiali-Istio-Checker [#[2145, 2146]](https://paste.fedoraproject.org/paste/fh~3LUgn6VwFrebaEnAw~A)
  * :heavy_check_mark: install-bookinfo [#[1173]](https://paste.fedoraproject.org/paste/V0lUAOMmwUNhX5AGqHwy2g)
  * :heavy_check_mark: Kiali-on-Openshift [#[1171]](https://paste.fedoraproject.org/paste/77e~uT8mSCtAyCM1EaP5yQ)
  * :red_circle: run-kiali-e2e-tests [#[574]](https://paste.fedoraproject.org/paste/BbfBgDhg-L8FZCmBu3b5Iw)


KIALI_NAME=kialicorepr588-istio-system

rhqci commented:
:heavy_check_mark: Jenkins CI: kiali-core-pr-e2e-test [#282](https://paste.fedoraproject.org/paste/LTUaLq9M9jyPRpyR24InEw)
  * :heavy_check_mark: Istio-and-Kiali-on-Openshift [#[1813]](https://paste.fedoraproject.org/paste/0jGGK3MgXFfAERGH-POB9Q)
  * :heavy_check_mark: Istio-Upstream [#[654]](https://paste.fedoraproject.org/paste/IYTC2U7GZIDkd5A2lCb1gw)
  * :heavy_check_mark: Kiali-Istio-Checker [#[2176, 2177]](https://paste.fedoraproject.org/paste/Or4plOUxYujh0~q4lVHlDw)
  * :heavy_check_mark: install-bookinfo [#[1192]](https://paste.fedoraproject.org/paste/HF-86Rw-LZ3ynyBlatSFvg)
  * :heavy_check_mark: Kiali-on-Openshift [#[1187]](https://paste.fedoraproject.org/paste/0yWELEBPZFLysQ-X~VpvDA)
  * :heavy_check_mark: run-kiali-e2e-tests [#[585]](https://paste.fedoraproject.org/paste/EqXhjEbVV0i39UG3RcWvTg)


KIALI_NAME=kialicorepr588-istio-system

lucasponce commented:
@hhovsepy mmm now the e2e tests seem to work but I can't make it work locally

hhovsepy commented:
@lucasponce  let me retest this PR

hhovsepy commented:
kialiqeci

rhqci commented:
:heavy_check_mark: Jenkins CI: kiali-core-pr-e2e-test [#286](https://paste.fedoraproject.org/paste/mcP1pOJn2le3L9y6aC99~Q)
  * :heavy_check_mark: Istio-and-Kiali-on-Openshift [#[1818]](https://paste.fedoraproject.org/paste/Q4-9I6UwLwDlYOObrMFsFQ)
  * :heavy_check_mark: Istio-Upstream [#[659]](https://paste.fedoraproject.org/paste/1Y1t5Fz0z4~7vCOi8UDpBQ)
  * :heavy_check_mark: Kiali-Istio-Checker [#[2187, 2188]](https://paste.fedoraproject.org/paste/IlHrYm2NrdDnHQr0x-IrEA)
  * :heavy_check_mark: install-bookinfo [#[1198]](https://paste.fedoraproject.org/paste/43Lj8f4yicQTP2gNWDwfaA)
  * :heavy_check_mark: Kiali-on-Openshift [#[1193]](https://paste.fedoraproject.org/paste/NlgxtT3qBojw2lDtg~LZ-g)
  * :heavy_check_mark: run-kiali-e2e-tests [#[595]](https://paste.fedoraproject.org/paste/9mKthqFgSObb0jnSSDr94A)


KIALI_NAME=kialicorepr588-istio-system

lucasponce commented:
@jotak @jmazzitelli Please, final look at this PR. I would like to merge it before it hits conflicts.

lucasponce commented:
Mmm the verbose it doesn't work well, I need to up the global verbose to 3

lucasponce commented:
@jmazzitelli moved globally to 3

rhqci commented:
:heavy_check_mark: Jenkins CI: kiali-core-pr-e2e-test [#289](https://paste.fedoraproject.org/paste/8l~U8rZ03ROFfnm9N9aPNQ)
  * :heavy_check_mark: Istio-and-Kiali-on-Openshift [#[1821]](https://paste.fedoraproject.org/paste/FSZktNgFsCoQkh3ijbmwcA)
  * :heavy_check_mark: Istio-Upstream [#[665]](https://paste.fedoraproject.org/paste/qrqGxzg2QCUMhiarfQlApw)
  * :heavy_check_mark: Kiali-Istio-Checker [#[2197, 2198]](https://paste.fedoraproject.org/paste/TbtBDDwl0e45dppgpMqsqw)
  * :heavy_check_mark: install-bookinfo [#[1201]](https://paste.fedoraproject.org/paste/wg~pFEo~UBZM68FpNGjVnA)
  * :heavy_check_mark: Kiali-on-Openshift [#[1198]](https://paste.fedoraproject.org/paste/gryx6jjZTa-mjiRf54EHFw)
  * :heavy_check_mark: run-kiali-e2e-tests [#[598]](https://paste.fedoraproject.org/paste/CuX~f2bFelzdhaWkV0yEmQ)


KIALI_NAME=kialicorepr588-istio-system

rhqci commented:
:heavy_check_mark: Jenkins CI: kiali-core-pr-e2e-test [#290](https://paste.fedoraproject.org/paste/xknGElk3NKytG0IYhSv7sA)
  * :heavy_check_mark: Istio-and-Kiali-on-Openshift [#[1822]](https://paste.fedoraproject.org/paste/q86VHpfdRdwHjFQfmELkqg)
  * :heavy_check_mark: Istio-Upstream [#[666]](https://paste.fedoraproject.org/paste/aVLRA694-jf91Fn1~KlnwQ)
  * :heavy_check_mark: Kiali-Istio-Checker [#[2200, 2199]](https://paste.fedoraproject.org/paste/slt0tHcRqlevGp6LFU7mlg)
  * :heavy_check_mark: install-bookinfo [#[1202]](https://paste.fedoraproject.org/paste/rOxbQzzV64tzVF~9o7OqvQ)
  * :heavy_check_mark: Kiali-on-Openshift [#[1199]](https://paste.fedoraproject.org/paste/X3Ln1soZHk~qGE7js3gJQA)
  * :heavy_check_mark: run-kiali-e2e-tests [#[599]](https://paste.fedoraproject.org/paste/v7fB-l5ZmmOsJucAFfXEgA)


KIALI_NAME=kialicorepr588-istio-system

rhqci commented:
:heavy_check_mark: Jenkins CI: kiali-core-pr-e2e-test [#291](https://paste.fedoraproject.org/paste/6TbhLc-64EKAOfVvDwmCzg)
  * :heavy_check_mark: Istio-and-Kiali-on-Openshift [#[1823]](https://paste.fedoraproject.org/paste/0ambiJUexby8anWsXjuj2w)
  * :heavy_check_mark: Istio-Upstream [#[667]](https://paste.fedoraproject.org/paste/IqFSUTINwJSatu5kD1h~Kw)
  * :heavy_check_mark: Kiali-Istio-Checker [#[2201, 2202]](https://paste.fedoraproject.org/paste/hKmODurQOpoFtjczYGNybw)
  * :heavy_check_mark: install-bookinfo [#[1203]](https://paste.fedoraproject.org/paste/b8CVaIz1xPlUKbOcimvmSw)
  * :heavy_check_mark: Kiali-on-Openshift [#[1200]](https://paste.fedoraproject.org/paste/GSbAR0SCME2DmRJmYwKMlg)
  * :heavy_check_mark: run-kiali-e2e-tests [#[600]](https://paste.fedoraproject.org/paste/H5QWYChcIMKzTf6xnNwWXg)


KIALI_NAME=kialicorepr588-istio-system

jotak commented on kubernetes/cache.go:
```diff
@@ -0,0 +1,458 @@
+		syncCount:       0,
+		maxSyncCount:    20, // Move this to config ? or this constant is good enough ?
+	}
+	registerErrorCallback(newControllerImpl.ErrorCallback)
```
I think we should be able to unregister error callbacks when cache controller is stopped, no?
So, as they are stored in a global slice, maybe "register" could return the index in slice, and on "stop" it would remove from slice using that index?

jotak replied:
Actually when I write "when cache controller is stopped" I'm not sure about what "stop" means here. What I want to say is, when the controller reaches its end of life.

lucasponce replied:
+1
But note that this registration happens once per IstioClient, which has not a destroy/clean method an relies on the lifecycle of the app.
This can be better improved, but I guess that probably it can go in a following PR as we'll need to add the caches for istio objects, too.

jotak replied:
=> https://issues.jboss.org/browse/KIALI-1841

jotak commented on kubernetes/cache.go:
```diff
@@ -0,0 +1,458 @@
+			callback(err)
+		}
+	}
+	utilruntime.ErrorHandlers = customErrorHandler
```
It could be just:
```
utilruntime.ErrorHandlers = append(utilruntime.ErrorHandlers, func(err error) {
		for _, callback := range errorCallbacks {
			callback(err)
		}
	}
}
```

no?

rhqci commented:
:heavy_check_mark: Jenkins CI: kiali-core-pr-e2e-test [#292](https://paste.fedoraproject.org/paste/qTTHfs9o9hMZqUPW1exRPw)
  * :heavy_check_mark: Istio-and-Kiali-on-Openshift [#[1826]](https://paste.fedoraproject.org/paste/1hQE8N6uwZ8v0LjWI49bFg)
  * :heavy_check_mark: Istio-Upstream [#[669]](https://paste.fedoraproject.org/paste/RZzRcpT~ibSzcilEukK7BA)
  * :heavy_check_mark: Kiali-Istio-Checker [#[2208, 2207]](https://paste.fedoraproject.org/paste/arNUEJWhyYVxQla5wPWAcw)
  * :heavy_check_mark: install-bookinfo [#[1206]](https://paste.fedoraproject.org/paste/bDOR8d7Z5qk-MkERP4hwPw)
  * :heavy_check_mark: Kiali-on-Openshift [#[1203]](https://paste.fedoraproject.org/paste/wQbD6AvgjElW1m73vNpQnw)
  * :heavy_check_mark: run-kiali-e2e-tests [#[605]](https://paste.fedoraproject.org/paste/upTuB4M1PU1EPmrV2QocwQ)


KIALI_NAME=kialicorepr588-istio-system

rhqci commented:
:heavy_check_mark: Jenkins CI: kiali-core-pr-e2e-test [#293](https://paste.fedoraproject.org/paste/ERhbAIAbbECswT8HYPDKXw)
  * :heavy_check_mark: Istio-and-Kiali-on-Openshift [#[1828]](https://paste.fedoraproject.org/paste/L5QODGBuHuyqzp0ldB9mVw)
  * :heavy_check_mark: Istio-Upstream [#[671]](https://paste.fedoraproject.org/paste/7RzcQ7H3oSJupuDoD2bYKw)
  * :heavy_check_mark: Kiali-Istio-Checker [#[2212, 2213]](https://paste.fedoraproject.org/paste/jEpGifT6DFuegzrBm4dFtw)
  * :heavy_check_mark: install-bookinfo [#[1208]](https://paste.fedoraproject.org/paste/m58xUIGZWCe4l-uoQOSEbQ)
  * :heavy_check_mark: Kiali-on-Openshift [#[1205]](https://paste.fedoraproject.org/paste/z8e0JT-HN3Dx~MTe2hOzPQ)
  * :heavy_check_mark: run-kiali-e2e-tests [#[608]](https://paste.fedoraproject.org/paste/HitH9RevaHL7nh2CLAgqTg)


KIALI_NAME=kialicorepr588-istio-system

jmazzitelli commented on server/server.go:
```diff
@@ -2,6 +2,7 @@ package server
 
 import (
 	"fmt"
+	"github.com/kiali/kiali/business"
```
put this import below in the section with the rest of the kiali imports to match coding style

jmazzitelli commented on kubernetes/cache.go:
```diff
@@ -0,0 +1,458 @@
+	"sync"
+	"time"
+
+	"github.com/kiali/kiali/log"
```
this import should be below the third-party imports in the "kiali import section" to match coding style

jmazzitelli commented on kubernetes/cache.go:
```diff
@@ -0,0 +1,458 @@
+
+func (c *controllerImpl) HasSynced() bool {
+	if c.syncCount > c.maxSyncCount {
+		log.Errorf("Max attempts reached syncing cache. Error connecting to k8s API ")
```
I think you should log the c.syncCount here. This is using Errorf and I see a blank space after "API " in the message, so maybe you forgot? I would like to see this:

`log.Errorf("Max attempts reached syncing cache. Error connecting to k8s API: %v > %v", c.syncCount, c.maxSyncCount)`

jmazzitelli commented on kubernetes/cache.go:
```diff
@@ -0,0 +1,458 @@
+	c.lastErrorLock.Lock()
+	if c.isErrorState == true {
+		// ping to check if backend is still unavailable (used namespace endpoint)
+		_, err := c.clientset.CoreV1().Namespaces().List(emptyListOptions)
```
Is this the least expensive API call we could make? Rather than ask for the list of namespaces, is there some kind of "status" endpoint? Or something that has a more predictably quick response time?

lucasponce replied:
This is a one time operation to recover full cache when backend is down.
clientset.Interface doesn't map a generic status function, so I used CoreV1() as representative here.
Fetching namespaces sounds correct to check that API is good.

jmazzitelli commented on kubernetes/cache_test.go:
```diff
@@ -0,0 +1,166 @@
+package kubernetes
+
+import (
+	"github.com/stretchr/testify/assert"
```
this github.com third party lib import should be in the second section of the imports to match coding style.

jmazzitelli commented on kubernetes/client.go:
```diff
@@ -79,6 +79,10 @@ type IstioClient struct {
 	k8s                *kube.Clientset
 	istioConfigApi     *rest.RESTClient
 	istioNetworkingApi *rest.RESTClient
+	// isOpenShift can be cached per client
```
Again, I would like to see a comment here to explicitly explain to the user why this is a pointer - we need to explain to devs looking at this code that it is expected and a very valid value that this variable be nil (when isOpenShift has yet to be determined).

jmazzitelli commented on kubernetes/client.go:
```diff
@@ -135,13 +144,24 @@ func NewClient() (*IstioClient, error) {
 // It returns an error on any problem.
 func NewClientFromConfig(config *rest.Config) (*IstioClient, error) {
 	client := IstioClient{}
+	log.Infof("Rest perf config QPS: %f Burst: %d", config.QPS, config.Burst)
```
is there a reason QPS is logged with floating point precision %f but Burst is not? I've only seen these settings being set to integer values (e.g. 175, or 200, or whatever).

lucasponce replied:
QPS is represented as a float and Burst is an integer.
It sounded logical to use %f and %d here.

rhqci commented:
:heavy_check_mark: Jenkins CI: kiali-core-pr-e2e-test [#295](https://paste.fedoraproject.org/paste/G12Py~4DwWcAz3sAVX9bsg)
  * :heavy_check_mark: Istio-and-Kiali-on-Openshift [#[1830]](https://paste.fedoraproject.org/paste/hyTpulMT~lKxonh0oHutZg)
  * :heavy_check_mark: Istio-Upstream [#[673]](https://paste.fedoraproject.org/paste/DVu0n5CUFnCtoPRYypN7uA)
  * :heavy_check_mark: Kiali-Istio-Checker [#[2218, 2219]](https://paste.fedoraproject.org/paste/MMEkZji6-3ybbzs~QX4BHg)
  * :heavy_check_mark: install-bookinfo [#[1210]](https://paste.fedoraproject.org/paste/-wdksI1E7FYrnBuigpiqtQ)
  * :heavy_check_mark: Kiali-on-Openshift [#[1208]](https://paste.fedoraproject.org/paste/l4I6oRLlKsgAECsZI2dwJA)
  * :heavy_check_mark: run-kiali-e2e-tests [#[612]](https://paste.fedoraproject.org/paste/x7poQysrq0gRhy65GUYsug)


KIALI_NAME=kialicorepr588-istio-system

rhqci commented:
:heavy_check_mark: Jenkins CI: kiali-core-pr-e2e-test [#296](https://paste.fedoraproject.org/paste/0kI1~6eZrFtdmG~3YQgK4w)
  * :red_circle: Istio-and-Kiali-on-Openshift [#[1831]](https://paste.fedoraproject.org/paste/5knGRZCgGIbrsmd-XMrc3Q)
  * :heavy_check_mark: Istio-Upstream [#[674]](https://paste.fedoraproject.org/paste/enGUI53s-GoflIkjty2KGg)
  * :heavy_check_mark: Kiali-Istio-Checker [#[2220]](https://paste.fedoraproject.org/paste/MQ67MfEx2ILNVS5MMlTdNg)
  * :heavy_check_mark: install-bookinfo [#[1211]](https://paste.fedoraproject.org/paste/Ique8VuOIc-kD-o6pRheRw)
  * :red_circle: Kiali-on-Openshift [#[1209]](https://paste.fedoraproject.org/paste/1MEzgv8y~zUn7VJyB8fzNQ)


KIALI_NAME=kialicorepr588-istio-system

rhqci commented:
:heavy_check_mark: Jenkins CI: kiali-core-pr-e2e-test [#297](https://paste.fedoraproject.org/paste/yR5X8tMdBFFIN5FNWdmEfQ)
  * :red_circle: Istio-and-Kiali-on-Openshift [#[1832]](https://paste.fedoraproject.org/paste/NYYUWxH57SqmpOc6f152iQ)
  * :heavy_check_mark: Istio-Upstream [#[675]](https://paste.fedoraproject.org/paste/PYW7HIWXjpAYhEuLCqh4dQ)
  * :heavy_check_mark: Kiali-Istio-Checker [#[2221]](https://paste.fedoraproject.org/paste/Kiu-CB3GslGU53IyjckBKg)
  * :heavy_check_mark: install-bookinfo [#[1212]](https://paste.fedoraproject.org/paste/ueqCQVA2~G6KEboVIZatpg)
  * :red_circle: Kiali-on-Openshift [#[1210]](https://paste.fedoraproject.org/paste/zZRVapOKk5cJNykZ9-OHeg)


KIALI_NAME=kialicorepr588-istio-system
